File: 2019/01/day1.hs
import System.IO  
import Text.Printf

main :: IO ()
main = do
  content <- getContents
  let input = map read (lines content)
  printf "Part 1: %d\n" (part1 input)
  printf "Part 2: %d\n" (part2 input)

part1 :: [Integer] -> Integer
part1 xs = sum (map cost xs)

part2 :: [Integer] -> Integer
part2 xs = sum (map f xs) - sum xs where
    f x | x > 0 = x + f (cost x)
        | otherwise = 0

cost :: Integer -> Integer
cost x = div x 3 - 2

File: 2019/01/day1.kt
package moe.kageru.aoc

fun main() {
    val input = generateSequence(::readLine)
        .map { it.toLong() }
        .toList();
    println("Part 1: ${part1(input)}")
    println("Part 2: ${part2(input)}")
}

fun part1(numbers: List<Long>): Long = numbers.map(::cost).sum()

fun part2(numbers: List<Long>): Long = numbers.map { costRec(it, 0) }.sum()

fun cost(mass: Long): Long = (mass / 3) - 2

tailrec fun costRec(mass: Long, acc: Long): Long {
    val cost = cost(mass)
    return when {
        cost <= 0 -> acc
        else -> costRec(cost, acc + cost)
    }
}
